### DISCUSSION TITLE
Create entity linked to another entity by just passing the foreign key

### DISCUSSION THREAD

**Discussion by trompx** [mikro-orm/mikro-orm#1499] [Q&A]
Hello,

I was trying to create an entity which is linked to another via ManyToOne.
I'm used to do raw queries, my guess was that I needed to just pass the foreign key of the linked entity of B to create A but it didn't work.

Base entity:
```ts
import { ID, Field, InterfaceType } from "type-graphql";
import { PrimaryKey } from "@mikro-orm/core";

@InterfaceType()
export abstract class BaseEntity {
  @Field(() => ID)
  @PrimaryKey()
  id!: string;
}
```

Entity A:
```ts
import {
  Entity,
  ManyToOne,
} from "@mikro-orm/core";
import { ObjectType, Field } from "type-graphql";
import { BaseEntity } from "../base/base.entity";
import { B } from "modules/entities";

@ObjectType({ implements: BaseEntity })
@Entity()
export class A extends BaseEntity {
  @Field()
  @Property()
  name: string;

  @Field(() => B)
  @ManyToOne("B")
  b: B;
}
```

Entity B:
```ts
import {
  Entity,
  OneToMany,
  Collection
} from "@mikro-orm/core";
import { ObjectType, Field } from "type-graphql";
import { BaseEntity } from "../base/base.entity";
import { A } from "modules/entities";

@ObjectType({ implements: BaseEntity })
@Entity()
export class B extends BaseEntity {
  @Field()
  @Property()
  name: string;

  @Field(() => [A], { nullable: true })
  @OneToMany("A", "b")
  a = new Collection<A>(this);
}
```

Tried to create with:
```ts
    const a = await this.repository.create({
      b: req.session.bId,
      ...data,
    });
```
But then I get the error `[GraphQL] Cannot return null for non-nullable field B.name.`
However, the following works:
```ts
    const b = await this.bRepository.findOne(req.session.bId);
    const a = await this.repository.create({
      b,
      ...data,
    });
```
Problem is I need to do an **additional query** each time I want to create a...
Is there a way to just pass the **bId** instead of the **b whole entity**?

Another weird thing, even doing this works:
```ts
    const b = await this.bRepository.findOne(req.session.bId);
    const a = await this.repository.create({
      // Just using bId, I manage to create a without error (without using b queried above)
      b: req.session.bId,
      ...data,
    });
```
Any idea why this works (without passing b that I just created and still use just the bId)?

**Comment by B4nan** [comment]
`em.create` is sync method, all it does is to create entity instance, so no need to await it - but you need to persist and flush. 

https://stackoverflow.com/questions/66386900/are-entities-created-with-em-create-managed-and-persisted

You don't need to query the actual entity just because you need its PK, providing PK is enough in `em.create`. You can also use `em.getReference()` which will create entity reference = entity with just the PK, without making no queries. If you pass just FK to `em.create`, it will call the `em.getReference()` for you behind the scenes.

Also note that thanks to identity map, `em.getReference()` will give you the already loaded entity if you did load it before - there is always only one instance and `em.getReference()` respects that.

```ts
const b = em.getReference(B, bid);
```

Without seeing full code, its hard to say what you got wrong. But note that your error is from GQL and probably has nothing to do with persisting the entity as is - enable debug mode and you will probably see the ORM part actually works fine. This sounds more like you are actually trying to read the `b.name`, but if you don't load the entity, there is no `b.name`.

**Reply by trompx** [reply]
Thanks a lot for taking the time @B4nan!

I'm aware it is sync method, here I don't call em.create, but `this.repository.create` which in the mikro ORM repository call em.create. In that case I could remove it but I could change ORM or what I return in my create method in the future, so I have set it in the repository like `async create`.

The full setup:
```ts
export class A extends BaseEntity {
  @Property()
  name: string;

  @ManyToOne("B")
  b: B;

  @OneToMany("C", "a")
  c = new Collection<C>(this);
}

@ObjectType()
export class CreateAResponse {
  @Field(() => A, { nullable: true })
  a?: A;

  // NOT WORKING
  // @Field(() => CreateAPartialResponse, { nullable: true })
  // a?: CreateAPartialResponse;

  @Field(() => [FieldError], { nullable: true })
  errors?: FieldError[];
}

@ObjectType()
export class CreateAPartialResponse implements Partial<A> {
  @Property()
  name: string;

  // Here I cannot set b as it implements A and get error : Property 'b' in type 'CreateAPartialResponse' is not assignable to the same property in base type 'Partial<A>'.
  // I cannot set bId: string as I get the error : Property 'bId' is missing in type 'A', even if I remove implements Partial<A> from the CreateAPartialResponse
  @Property()
  b: string;

  // Here if I repeat the c field setup that I have in A, I get the error :  error TS2322: Type 'A & { c: LoadedCollection<C, C>; }' is not assignable to type 'CreateAPartialResponse | undefined'.
  // If I remove implements Partial<A> from the CreateAPartialResponse, and either (keep declaring c = new...) OR (completely removing it from the response) I still get the error TS2322
  @OneToMany("C", "a")
  c = new Collection<C>(this);
}
```

You were right. In fact doing this throws the error but still manage to create a in DB.
```ts
  async create(
    data: AInput,
    req: any
  ): Promise<CreateAResponse> {
    const a = await this.repository.create({
      b: req.session.bId,
      ...data,
    });
    ... (handle if error)
    return { a };
  }
```
I guess the problem as you said is that here I return a promise telling that it will return CreateAResponse, and as CreateAResponse has `a?: A` which has one `b: B` and here **instead of returning a whole instance of b, I only return bId**, I get the mentionned error.

I tried to create a CreateAPartialResponse partial response like described above but I couldn't make the setup work.
Any idea how I could setup the partial response so I can just return the a.name, a.bId OR a.b.id and a.c = null?

**Reply by B4nan** [reply]
i don't really understand what you are up to, but if this is really "full code", then you are missing `@Entity()` decorator and the rest does not make sense at all - how can you use MikroORM entity and typegraphql decorators? what would that mean? MikroORM decorators should be used only on entities and I don't understand how could just implementing an `Partial<A>` have any meaning (it is just a compile time check, interfaces are removed).

**Reply by trompx** [reply]
My bad @B4nan, an sorry for the previous post, I totally messed up when I trimmed my code.

Here is what I have:
```ts
import {
  Entity,
  ManyToOne,
  OneToMany,
  Collection
} from "@mikro-orm/core";
import { ID, Field, ObjectType, InterfaceType } from "type-graphql";
import { PrimaryKey } from "@mikro-orm/core";

@InterfaceType()
export abstract class BaseEntity {
  @Field(() => ID)
  @PrimaryKey()
  id!: string;
}

@ObjectType({ implements: BaseEntity })
@Entity()
export class List extends BaseEntity {
  @Field()
  @Property()
  name: string;

  @Field(() => User)
  @ManyToOne("User")
  user: User;

  @OneToMany("Item", "list")
  items = new Collection<Item>(this);
}

@ObjectType({ implements: BaseEntity })
@Entity()
export class User extends BaseEntity {
  @Field()
  @Property()
  name: string;

  @Field(() => [List], { nullable: true })
  @OneToMany("List", "user")
  lists = new Collection<List>(this);
}

@ObjectType({ implements: BaseEntity })
@Entity()
export class Item extends BaseEntity {
  @Field()
  @Property()
  name: string;

  @Field(() => List)
  @ManyToOne("List")
  list: List;
}

@ObjectType()
export class CreateListResponse {
  @Field(() => CreateListPartialResponse, { nullable: true })
  list?: CreateListPartialResponse;

  @Field(() => [FieldError], { nullable: true })
  errors?: FieldError[];
}

// Here I try to create a response so when I create a list in the client, I don't have a graphql error telling that I'm missing user data (like it's name) as I'm just setting the userId.
@ObjectType()
export class CreateListPartialResponse {
  @Field()
  name: string;

  @Field()
  userId: string;

  @Field(() => [Item], { nullable: true })
  items: Item[];
}
```

And the error I get when trying to create a list:
```ts
  async create(
    data: ListInput,
    req: any
  ): Promise<CreateListResponse> {
    const list = await this.repository.create({
      userId: req.session.userId,
      ...data,
    });

    // I get two errors:
    // Type 'List & { items: LoadedCollection<Item, Item>; }' is not assignable to type 'CreateListPartialResponse | undefined'.
    // Property 'userId' is missing in type 'List & { items: LoadedCollection<Item, Item>; }' but required in type 'CreateListPartialResponse'.ts(2322)
    return { list };
  }
```
I don't know how to shape the CreateListPartialResponse so I can return the new list created (which doesn't have a full user entity, and have no initial items) without a graphql error.
I hope this example makes more sense. Let me know if I can provide more information.
